routing/router.go

```diff
@@ -2293,3 +2293,278 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
 
 	return bandwidthHints, nil
 }
+
+// runningAmounts keeps running amounts while the route is traversed.
+type runningAmounts struct {
+	// amt is the intended amount to send via the route.
+	amt lnwire.MilliSatoshi
+
+	// max is the running maximum that the route can carry.
+	max lnwire.MilliSatoshi
+}
+
+// checkOrUpdateAmt checks or updates the running amounts while
+// traversing the channels.
+func (r *ChannelRouter) checkOrUpdateAmt(policy *channeldb.ChannelEdgePolicy,
+	capacity btcutil.Amount, amts runningAmounts, localChan bool,
+	useMinAmt bool) (runningAmounts, error) {
+
+	// Determine max htlc value.
+	maxHtlc := lnwire.NewMSatFromSatoshis(capacity)
+	if policy.MaxHTLC != 0 {
+		maxHtlc = policy.MaxHTLC
+	}
+
+	amt := amts.amt
+
+	// If we have a specific amount for which we are building the route,
+	// validate it against the channel constraints and return the new
+	// running amount.
+	if !useMinAmt {
+		if amt < policy.MinHTLC || amt > maxHtlc {
+			return runningAmounts{}, fmt.Errorf("channel htlc "+
+				"constraints [%v - %v] violated with amt %v",
+				policy.MinHTLC, maxHtlc, amt)
+		}
+
+		// Update running amount by adding the fee for non-local
+		// channels.
+		if !localChan {
+			amt += policy.ComputeFee(amt)
+		}
+
+		return runningAmounts{
+			amt: amt,
+		}, nil
+	}
+
+	// Adapt the minimum amount to what this channel allows.
+	if policy.MinHTLC > amts.amt {
+		amt = policy.MinHTLC
+	}
+
+	// Update the maximum amount too to be able to detect incompatible
+	// channels.
+	max := amts.max
+	if policy.MaxHTLC < amts.max {
+		max = policy.MaxHTLC
+	}
+
+	// If we get in the situation that the minimum amount exceeds
+	// the maximum amount (enforced further down stream), we have
+	// incompatible channel policies.
+	//
+	// There is possibility with pubkey addressing that we should
+	// have selected a different channel downstream, but we don't
+	// backtrack to try to fix that.
+	if amt > max {
+		return runningAmounts{}, errors.New("incompatible channel policies")
+	}
+
+	// Add fees to the running amounts. Skip the source node fees as
+	// those do not need to be paid.
+	if !localChan {
+		amt += policy.ComputeFee(amt)
+		max += policy.ComputeFee(max)
+	}
+
+	return runningAmounts{amt: amt, max: max}, nil
+}
+
+// ErrNoChannel is returned when a route cannot be built because there are no
+// channels that satisfy all requirements.
+type ErrNoChannel struct {
+	position int
+	fromNode route.Vertex
+}
+
+func (e ErrNoChannel) Error() string {
+	return fmt.Sprintf("no matching outgoing channel available for "+
+		"node %v (%v)", e.position, e.fromNode)
+}
+
+// BuildRoute returns a fully specified route based on a list of pubkeys. If
+// amount is nil, the minimum routable amount is used. To force a specific
+// outgoing channel, use the outgoingChan parameter.
+func (r *ChannelRouter) BuildRoute(amt *lnwire.MilliSatoshi,
```

Comment:
is method strictly dependent on the router, or could it be defined in a smaller file? 😛

Reply:
I don't think it gets very pretty if this method would be moved to a new struct. If we want to do something about the size of `ChannelRouter`, it probably makes more sense to separate the graph update part from the routing part.